x/ccv/provider/keeper/partial_set_security.go

```diff
@@ -67,3 +72,152 @@ func (k Keeper) HandleOptOut(ctx sdk.Context, chainID string, providerAddr types
 
 	return nil
 }
+
+// ComputeValidatorUpdates computes the validator updates needed to be sent to the consumer chain to capture
+// the newly opted-in and opted-out validators, as well as validators that unbonded.
+func (k Keeper) ComputeValidatorUpdates(ctx sdk.Context,
+	currentValidators []OptedInValidator,
+	validatorAddressesToAdd []types.ProviderConsAddress,
+	validatorAddressesToRemove []types.ProviderConsAddress,
+) []abci.ValidatorUpdate {
+	var m = make(map[string]abci.ValidatorUpdate)
+
+	for _, val := range currentValidators {
+		validator, found := k.stakingKeeper.GetValidatorByConsAddr(ctx, val.ProviderAddr.ToSdkConsAddr())
+		if !found {
+			continue
+		}
+
+		pubKey, err := validator.TmConsPublicKey()
+		if err != nil {
+			continue
+		}
+
+		// if the voting power did not change since the last time the validator opted in, do not create an update
```

Comment:
The comment should probably be rephrased to:
```
// if the voting power is the same as the one the validator had when it opted in, do not create an update
```
because the voting power could change multiple times during an epoch, but at the end of an epoch, it could be that the validator has again the same voting power it had when it opted in.

> Also, to better understand, with epochs, GetLastValidatorPower will need to be replaced with something like GetValidatorPowerAtLastEpoch, right?

Not sure I understand what you mean here. `GetLastValidatorPower` returns the most recent validator power and `ComputeValidatorUpdates` would be called at the end of every epoch.
Think of the following example. A validator `V` opts in at the end of an epoch `e` with power `p`. If at a later  epoch `e + X` (`X >= 1`), validator `V` has still the same power `p`, then we should not send a `ValidatorUpdate` for `V` and this is what this checks here.

Reply:
Right. GetLastValidatorPower will return the validator power in the last block. But what we need to compare against is the validator power at the last epoch. If epochLength==1, (as is the case right now) these coincide, but if epochLength != 1 this is not the case anymore.